x-pack/plugins/observability/public/context/has_data_context.tsx

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { isEmpty } from 'lodash';
+import React, { createContext, useEffect, useState } from 'react';
+import { getDataHandler } from '../data_handler';
+import { FETCH_STATUS } from '../hooks/use_fetcher';
+import { useRouteParams } from '../hooks/use_route_params';
+import { useTimeRange } from '../hooks/use_time_range';
+import { ObservabilityFetchDataPlugins, UXHasDataResponse } from '../typings/fetch_overview_data';
+
+export type HasDataMap = Record<
+  ObservabilityFetchDataPlugins,
+  { status: FETCH_STATUS; hasData?: boolean | UXHasDataResponse }
+>;
+
+export interface HasDataContextValue {
+  hasData: Partial<HasDataMap>;
+  hasAnyData: boolean | undefined;
+}
+
+export const HasDataContext = createContext({} as HasDataContextValue);
+
+const apps: ObservabilityFetchDataPlugins[] = [
+  'apm',
+  'uptime',
+  'infra_logs',
+  'infra_metrics',
+  'ux',
+];
+
+export function HasDataContextProvider({ children }: { children: React.ReactNode }) {
+  const { rangeFrom, rangeTo } = useRouteParams('/overview').query;
+  const { absStart, absEnd } = useTimeRange({ rangeFrom, rangeTo });
+
+  const [hasData, setHasData] = useState<HasDataContextValue['hasData']>({});
+
+  useEffect(
+    () => {
+      apps.forEach(async (app) => {
+        try {
+          const params =
```

Comment:
Why does the UX app have a different interface for its hasData handler? And looks like it will not be updated because it's not a dependency?

Reply:
I think @shahzad31 is the right person to answer this question.

Reply:
Maybe @justinkambic knows?

Reply:
@dgieselaar i believe it was a compromise between performance and consistency. For UX case we are displaying data for the service with most traffic, so in the same query where we are getting if it has rum data , it is also getting service name with most traffic via terms aggs. 

I believe we need better types here to clear this to avoid this if conditions. Or maybe we can get service name via a different query or maybe make two es queries in data request. IMO consistency here isn't worth an additional query. 

There was also discussion about this in original PR where this was added.

Reply:
Hmm, I would expect that to be unlikely. The fact that you're running a terms aggregation to get the service name means that you cannot terminate early, because all documents have to be collected and aggregated (within the given time range). That will make your `hasData` call significantly slower. I don't think it's worth it, in fact, it's probably worse in terms of performance. Instead of running things in parallel, and getting certain parts of the data when they're ready, it's now one long blocking request.

Reply:
with this PR, it won't be blocking, i mean data will be loaded as soon as any has data call is resolved. this is why i wanted this to be done. 

Anyway i am not against a refactor.

Reply:
btw i am not sure terminate early even provides any benefit here, the call on dev-next isn't bad compared to others. uptime also doesn't use that flag, and it's fastest. 


![image](https://user-images.githubusercontent.com/3505601/96589423-155a9200-12e5-11eb-9943-e87fc2a60c2e.png)

Reply:
i know this is just one call and doesn't necessarily reflects the performance benchmark.

Reply:
it might be a case that terminate early flag isn't being passed correctly to es request.

Reply:
dev-next doesn't have a lot of data. IIRC, half of our users have more data, and significantly less nodes. So, if it's slow on dev-next it's most likely an issue, but if it's fast, it doesn't really mean that it will be fast for our users. In this example, you'll likely have relatively high overhead of security checks etc, which happen for each request, so it's hard to tell how efficient the search is based on network timings.

I think it's good here to take a step back and analyze what's happening with this search:

```ts
{
  size: 0,
  query: {
    bool: {
      filter: [{ term: { [TRANSACTION_TYPE]: TRANSACTION_PAGE_LOAD } }],
    },
  },
  aggs: {
    services: {
      filter: {
        range: rangeFilter(start, end),
      },
      aggs: {
        mostTraffic: {
          terms: {
            field: SERVICE_NAME,
            size: 1,
          },
        },
      },
    },
  },
}
```

- First _all_ page load transactions in all indices are collected, because there is no range query on `@timestamp`.
- Then, a filter aggregation runs to select only documents that match the given time range.
- Then, another aggregation runs on all the documents for the given time range. `size` only removes the bucket from the response - it will still be calculated.

All these steps run sequentially (per document).

Especially the first step could be pretty slow for users that have a lot of RUM data.

If you separate this however, you'd have two searches that you can run in parallel, plus, even the total duration will be faster.

- For the `hasData` search, you can set `terminate_after` to 1, meaning it will exit as soon as it has found a document. The performance cost of this search is negligible in most cases.
- For the `getServiceName` search, you can simply add a range query in your bool clause. This will remove the need for a filter aggregation.

Reply:
One example from DPP:

- one dual-purpose search: 2000ms
- hasData: 40ms
- getServiceName: 90ms

Reply:
Queries:

```
GET apm-*/_search?request_cache=false
{
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "term": {
            "transaction.type": "page-load"
          }
        }
      ]
    }
  },
  "aggs": {
    "services": {
      "filter": {
        "range": {
          "@timestamp": {
            "gte": "2020-05-12T10:00:00.000Z",
            "lt": "2020-05-12T11:00:00.000Z"
          }
        }
      },
      "aggs": {
        "mostTraffic": {
          "terms": {
            "field": "service.name",
            "size": 1
          }
        }
      }
    }
  }
}

GET apm-*/_search?request_cache=false
{
  "query": {
    "bool": {
      "filter": [
        {
          "term": {
            "transaction.type": "page-load"
          }
        }
      ]
    }
  },
  "size": 0,
  "track_total_hits": true,
  "terminate_after": 1
}

GET apm-*/_search?request_cache=false
{
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "term": {
            "transaction.type": "page-load"
          }
        },
        {
          "range": {
            "@timestamp": {
              "gte": "2020-05-12T10:00:00.000Z",
              "lt": "2020-05-12T11:00:00.000Z"
            }
          }
        }
      ]
    }
  },
  "aggs": {
    "mostTraffic": {
      "terms": {
        "field": "service.name",
        "size": 1
      }
    }
  }
}
```

Reply:
++